storage2/src/state/transaction.rs

```diff
@@ -0,0 +1,43 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+
+use jmt::storage::{TreeReader, TreeWriter};
+
+use super::{State, StateWrite};
+
+/// Represents a transactional set of changes to a `State` fork,
+/// implemented as a RYW cache over a `State`.
+pub(crate) struct Transaction<'a, R: TreeReader + TreeWriter> {
+    // TODO: should higher-level types be used here instead of jmt types?
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
+    state: &'a State<'a, R>,
```

Comment:
We shouldn't fill in the internals here, because we haven't designed how it should work, and we can't do that (#1490) until we have an API sketch.

The storage `Transaction` shouldn't take a generic `TreeReader + TreeWriter`, since we aren't plugging in an arbitrary storage backend.

Having a bound on the lifetime of the `State` the transaction is modifying seems reasonable -- but perhaps we should be holding a `&'a mut State`, to enforce that there can only be one live `Transaction` at a time?